### PR TITLE
8242564: javadoc crashes:: class cast exception com.sun.tools.javac.code.Symtab$6

### PR DESCRIPTION
**Pull request by vpa1977**
'--ignore-source-errors' allows generating Javadoc for the packages that contain compilation errors. 

jdk.javadoc.internal.doclets.toolkit.util.ClassTree generates a type hierarchy for javadoc that may include error types such as 
```
class Foo extends Bar {
}
``` 
where Bar is undefined. 

The user still wants to generate documentation for Foo and have Bar as a text label. 

For the unknown class Bar it is impossible to detect the enclosing class/file and javadoc crashes with exception. 

This PR returns Kind.OTHER for the error types, avoiding the javadoc crash.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8242564](https://bugs.openjdk.org/browse/JDK-8242564): javadoc crashes:: class cast exception com.sun.tools.javac.code.Symtab$6 (**Bug** - P3)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17435/head:pull/17435` \
`$ git checkout pull/17435`

Update a local copy of the PR: \
`$ git checkout pull/17435` \
`$ git pull https://git.openjdk.org/jdk.git pull/17435/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17435`

View PR using the GUI difftool: \
`$ git pr show -t 17435`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17435.diff">https://git.openjdk.org/jdk/pull/17435.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17435#issuecomment-1894607437)

### CODE REVIEW

**Comment by vpa1977** [comment]
test ran at commit 53d29853fd703406247480b6a23976bb0693e5e1 fail due to the ClassCastException:
```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java
>>                                                       1     0     1     0 <<
==============================
TEST FAILURE
```

Test ran at the commit 9ebb1ac75dc3fe70133e434901ba858b391cf5d3 (tip of the branch(with fix included) pass:

```

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java
                                                         1     1     0     0   
==============================
TEST SUCCESS

Stopping javac server
Finished building target 'test' in configuration 'linux-x86_64-server-release'


$make TEST=:langtools_javadoc test

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/langtools:langtools_javadoc              500   500     0     0   
==============================
TEST SUCCESS
```

File: test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java [hunk 1]
```diff
@@ -43,6 +45,37 @@ public static void main(String... args) throws Exception {
...
+                "-sourcepath", testSrc,
+                "badpkg");
+        // Then javadoc exits successfully
+        checkExit(Exit.OK);
+        // And generates html for the ChildClass
+        checkOutput("badpkg/package-tree.html", true,
+        """
+        <li class="circle">badpkg.<a href="ChildClass.html" class="type-name-link" \
+        title="class in badpkg">ChildClass</a> (implements java.lang.Iterable&lt;T&gt;)</li>
+        """);
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java:66]: arguments, like this text block, should be indented.

In general, if you're using an IDE that can reformat fragments of code, like a method call, it should be enough/OK to use the IDE reformat calls like this.

File: test/langtools/jdk/javadoc/doclet/testClassTree/badpkg/ChildClass.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/*
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testClassTree/badpkg/ChildClass.java:1]: These days, the preferred style is to write small files on the fly during the test.  That way, the test class is better co-located with the the code that uses it, and as a secondary benefit, you do not need to include a full legal header that tends to overwhelm the few lines of interest.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/ToolEnvironment.java [hunk 3]
```diff
@@ -181,6 +182,8 @@ void setElementToTreePath(Element e, TreePath tree) {
     }
 
     public Kind getFileKind(TypeElement te) {
+        if (te.asType().getKind() == TypeKind.ERROR)
+            return Kind.OTHER;
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/ToolEnvironment.java:186]: On a medium deep-dive, I looked at whether this was the "best" solution, or whether it was "second best" and that we should fix the origin of the `ClassCastException` in `Symbol.outermostclass()`. Bottom line: while we could fix it for this case, in general, there is no way to guarantee that there is a reasonable result of type `ClassSymbol` for every reasonable symbol for which this method is called.  Arguably, `.outermostclass` could do more checking and/or throw some exception and/or return `null`, but that is not the javac style, and we would still have to deal with the issue at call sites, like this one.

It's also debatable whether the correct return value is `OTHER` or `null`. `OTHER` has typically meant _some kind of file with an unknown extension_ and not _no file_.  The spec is unclear on this point, and clarifying the spec is out of scope in this PR, and would require a CSR.  So, for now, `OTHER` is OK.

**Review by jonathan-gibbons — COMMENTED** [review]
Thanks for taking this on. 
The work does show up more problems than it solves (!!) such as the use of `JavaFileObject.Kind.OTHER`, and the inconsistent output generated by the test. But both of those are out of scope for this work, which is primarily about avoiding a crash caused by an unexpected exception.

There's a typo (`exeception`) in the title that I cannot fix for you, and there are some suggestions to improve the comments and class names in the test. Do those and you'll be good to go.

File: test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java [hunk 4]
```diff
@@ -23,26 +23,76 @@
...
     public static void main(String... args) throws Exception {
         var tester = new TestClassTree();
         tester.runTests();
     }
 
+    @Test
+    public void testBadPkg(Path base) throws IOException {
+        // Given badpkg package containing class ChildClass with an undefined
+        //       base class, implementing undefined interface and a defined
+        //       interface
```
[the remark this answers is not in the document]
Comment (vpa1977) [reply test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java:58]: Hi, those are a gherkin scenario[1] and to make things easier for a reader i can move it out into a javadoc comment for the method.


[1] https://cucumber.io/docs/gherkin/reference/

File: test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java [hunk 5]
```diff
@@ -23,26 +23,76 @@
...
+                    <li class="circle">badpkg.<a href="ChildClass.html" class="type-name-link" title="\
+                    class in badpkg">ChildClass</a> (implements java.lang.Iterable&lt;T&gt;)</li>
+                    """);
+        checkOutput("badpkg/ChildClass.html", true,
+                """
+                    <div class="type-signature"><span class="modifiers">public class </span>\
+                    <span class="element-name type-name-label">ChildClass</span>
+                    <span class="extends-implements">extends ParentClass
+                    implements java.lang.Iterable</span></div>
+                    """);
```
[the remark this answers is not in the document]
Comment (vpa1977) [reply test/langtools/jdk/javadoc/doclet/testClassTree/TestClassTree.java]: I've  split it into a separate issue - https://bugs.openjdk.org/browse/JDK-8324791, would it be ok for me to look into addressing it?

**Comment by vpa1977** [comment]
Thank you for reviewing PR! 
I've updated as per comments, and if you find that test comments are not strictly necessary, i can remove those.

**Review by jonathan-gibbons — APPROVED** [review]
Thanks for doing the final round of changes.

**Pull request closed without merging**